drafts/parsing-fixed-width-records-using-lenses.md

```diff
@@ -0,0 +1,333 @@
+---
+title: Efficiently Parsing COBOL Fixed-Length Records with Functional Lenses
+domain: software-engineering-corner.hashnode.dev
+tags: functional-programming,java,mainframe,cobol,ibm,legacy,legacy-systems,legacy-app-modernization
+cover: https://cdn.hashnode.com/res/hashnode/image/upload/v1725967073596/zuzeoPioj.jpg?auto=format
+publishAs: KevZ
+hideFromHashnodeCommunity: false
+saveAsDraft: true
+---
+
+I have been lucky enough to work on more than one project that required integration with legacy mainframe applications throughout my career at Zühlke.
+Integrating with legacy technologies can be challenging and requires innovative solutions to avoid exposing legacy abstractions through APIs and REST interfaces.
+In this particular article I would like to focus on one aspect, namely the parsing of fixed-width COBOL records.
+COBOL fixed-width records are a format used in data storage and processing, particularly on IBM mainframes, where each record (or row of data) has a predefined, fixed length.
+This means that every record in a file occupies the exact same number of bytes, and each field within a record has a fixed position and size.
+Unlike CSV (Comma-Separated Values) or other delimited formats, fixed-width records don't use characters like commas or tabs to separate fields.
+Let’s say you have a record that contains the following fields:
+
+- Employee ID (6 characters)
+- Employee Name (20 characters)
+- Date of Birth (8 characters)
+
+A fixed-width record for this could look like:
+
+```text
+000123Alice Doe           19840929
+000124Bob Doe             19840929
+```
+
+The employee ID is consistently 6 characters long, padded with leading zeros.
+Likewise, the employee name is fixed at 20 characters, filled with trailing spaces. 
+
+
+Parsing fixed width records require careful parsing since each field must be extracted based on its byte position, which can be cumbersome when the data schema changes.
+Over the years, I have seen various approaches and libraries trying to provide a friendly API and developer experience with varying success.
+Some libraries use annotations and mutable data structures to combine the schema with the data class.
+Other libraries separate the schema from the data class by introducing JSON or XML files.
+Personally, I don't like to use mutable data structures and  try to avoid them whenever I can.
```

Comment:
[Opinion]
Perhaps reference popular Java books that also preach immutability to show that it is a wide-spread understanding to improve code.

Eg: Effective Java, (Item 17)[https://learning.oreilly.com/library/view/effective-java-3rd/9780134686097/ch4.xhtml#lev17]